### PR TITLE
Calculate SNAP descriptors

### PR DESCRIPTION
**Pull request by jan-janssen**
This pull request introduces an interface to LAMMPS: https://docs.lammps.org/compute_sna_atom.html to calculate the SNAP descriptors directly from python. It primarily implements three functions: 
* `stk.analyse.get_snap_descriptor_names()` - to get the names of the SNAP components for a given 2j_max.
* `stk.analyse.calc_snap_descriptors_per_atom()` - to calculate the per atom SNAP descriptors, for example to analyse crystal defects. 
* `stk.analyse.calc_snap_descriptor_derivatives()` - to calculate the per atom SNAP descriptors and their derivatives to fit a machine learning potential.

### CODE REVIEW

File: structuretoolkit/analyse/snap.py
Comment (pmrv) [inline structuretoolkit/analyse/snap.py:98]: Maybe I'm not following the array shape closely enough, but it feels that some of the flattens are redundant.

File: structuretoolkit/analyse/snap.py
Comment (pmrv) [inline structuretoolkit/analyse/snap.py:191]: Is this not just `for k in [...]` since all the keys and values are equal?  I also see that those (and subsets of them) options are used later.  It'd be great to define them somewhere outside of the functions and give them names, because I don't follow at all right now.

File: structuretoolkit/analyse/snap.py
Comment (pmrv) [inline structuretoolkit/analyse/snap.py:332]: Seems like old debugging code, remove if not needed or raise an exception.

File: structuretoolkit/analyse/snap.py
Comment (pmrv) [inline structuretoolkit/analyse/snap.py:362]:
```suggestion
    )
```
Now I'm not sure anymore, but I'd have guessed that concatenate always gives a copy anyway.

File: structuretoolkit/analyse/snap.py
Comment (pmrv) [inline structuretoolkit/analyse/snap.py:379]: Am uneasy about silencing errors like this, but not sure if lammps even gives us something useful back.

File: structuretoolkit/analyse/snap.py
Comment (pmrv) [reply structuretoolkit/analyse/snap.py:31]: Oh, this actually not used anywhere?  Is it supposed to be user facing API?

**Review by pmrv — CHANGES_REQUESTED** [review]
Generally ok, but I left some nits.  I do want a clearer separation between which functions are API and which functions are helpers and better (or any) docstrings on the API ones.

**Comment by pmrv** [comment]
Oh, and I didn't understand the conda merge business.  This is needed to have lammps installed for the tests but not as a dep?  Can we not do it like in pyiron modules to have both environment files applied sequentially?

File: structuretoolkit/analyse/snap.py
Comment (liamhuber) [inline structuretoolkit/analyse/snap.py:154]: This is unnecessary as the entire innermost loop could instead just be truncated to start at `j1`

File: structuretoolkit/analyse/snap.py
Comment (liamhuber) [reply structuretoolkit/analyse/snap.py:45]: On the topic of hard to follow, what even is "apre"? This doesn't feel like it uses generic best practices on python variable naming

File: structuretoolkit/analyse/snap.py
Comment (liamhuber) [inline structuretoolkit/analyse/snap.py:118]: Why not just add the extra four characters to call it "pointer"? Since the docstring bothers mentioning about the memory I assume that's the justification for this name

File: structuretoolkit/analyse/snap.py
Comment (liamhuber) [inline structuretoolkit/analyse/snap.py:129]: For the workflow stuff we currently need a single return statement, but I've found that practice actually makes things a lot easier to read. Eg here if/elif/else, assigning something to a variable each time then return that variable after the loops. This makes it much more clear at a glance whether or not the number of return values are changing.

File: structuretoolkit/analyse/snap.py
Comment (liamhuber) [inline structuretoolkit/analyse/snap.py:138]: Imo these would all be much easier to read formatted something like

```python
for cmd in [
    "clear",
    ...
    "atom modify map array sort 0 2.0"
]:
    lmp.command(cmd)
```

The commands read even a bit more like a regular file and it avoids retyping `lmp.command` N times

**Review by liamhuber — COMMENTED** [review]
Not exhaustive or detailed, just some thoughts skimming through. Type hints would make it all easier to read.

File: structuretoolkit/analyse/snap.py
Comment (jan-janssen) [reply structuretoolkit/analyse/snap.py:36]: I added a docstrings and tests for the user facing functions and marked the internal functions as private functions.

File: structuretoolkit/analyse/snap.py
Comment (jan-janssen) [reply structuretoolkit/analyse/snap.py]: I removed the check.

File: structuretoolkit/analyse/snap.py
Comment (jan-janssen) [reply structuretoolkit/analyse/snap.py:515]: In general this is correct but here the entries of the numpy array are links to the values stored in the LAMMPS executable - to prevent duplicating the memory - so we have to copy them.

**Comment by jan-janssen** [comment]
> Oh, and I didn't understand the conda merge business. This is needed to have lammps installed for the tests but not as a dep? Can we not do it like in pyiron modules to have both environment files applied sequentially?

While it is technically possible to apply the files sequentially, this breaks the performance of the GitHub action. Still I agree that the `conda_merge.py` script is outdated and we can simply append the two environment files using `tail`.

File: structuretoolkit/analyse/snap.py
Comment (jan-janssen) [reply structuretoolkit/analyse/snap.py]: Done

File: structuretoolkit/analyse/snap.py
Comment (jan-janssen) [reply structuretoolkit/analyse/snap.py]: Done

File: structuretoolkit/analyse/snap.py
Comment (jan-janssen) [reply structuretoolkit/analyse/snap.py]: Done.

**Comment by jan-janssen** [comment]
> Generally ok, but I left some nits. I do want a clearer separation between which functions are API and which functions are helpers and better (or any) docstrings on the API ones.

I marked all the internal functions as private functions.

File: structuretoolkit/analyse/snap.py
Comment (jan-janssen) [reply structuretoolkit/analyse/snap.py]: Done

File: structuretoolkit/analyse/snap.py
Comment (jan-janssen) [reply structuretoolkit/analyse/snap.py:539]: Unfortunately LAMMPS just crashes hard, so we need this option when we want to iterate over multiple structures.

File: structuretoolkit/analyse/snap.py
Comment (jan-janssen) [reply structuretoolkit/analyse/snap.py]: This functionality is basically a reduced version of the [UnfoldingPrism](https://github.com/pyiron/pylammpsmpi/blob/main/pylammpsmpi/wrapper/ase.py#L376) in `pylammpsmpi`. But as `pylammpsmpi` is not available on windows I would prefer to not have it as a dependency for `structuretoolkit`.

File: structuretoolkit/analyse/snap.py
Comment (jan-janssen) [reply structuretoolkit/analyse/snap.py]: While I can understand the appeal, I do not think this is a general requirement, in particular for functionality like this which is more deeply nested inside.

**Comment by jan-janssen** [comment]
This pull request is again ready for review

File: structuretoolkit/analyse/snap.py
Comment (liamhuber) [inline structuretoolkit/analyse/snap.py:173]: I already commented on this in my last review.

```python
for j in range(j1 - j2, foo, bar):
    if j >= j1:
        do_something()
    else:
        do_nothing()
```

Is functionally equivalent to 

```
for j in range(j1, foo, bar):
        do_something()
```

Except that the former is both more code _and_ less efficient. Please fix this.

File: structuretoolkit/analyse/snap.py
Comment (liamhuber) [inline structuretoolkit/analyse/snap.py:474]: There exists a condition under which the [assert will get skipped](https://docs.python.org/3/reference/simple_stmts.html#the-assert-statement) -- are you sure you want this and not a try/except?

File: structuretoolkit/analyse/snap.py
Comment (liamhuber) [reply structuretoolkit/analyse/snap.py:539]: @pmrv's unease is extremely warranted though, so in circumstances like this where there's actually a good reason for silencing, I suggest to add a `#comment` to explain this to future devs

**Review by liamhuber — CHANGES_REQUESTED** [review]
Docstrings are in place, names make enough sense, and tests are in place and passing; I'm not super familiar with SNAP so I'm not going to dig further into details, and those things are enough to satisfy me. There remains a trivially inefficient for loop that I already complained about that needs to be fixed, but otherwise I have only the following non-blocking and details-agnostic suggestions for you to consider:

- Per my original review comment, the functions would benefit from type hints. Additionally, some of the docstring type hints can be expanded, e.g. is `atom_types (list):` actually `atom_types (list[str]):`
- I like that the function names have verbs in them, but the difference between `calc` and `get` is implicit. If `get` functions are cheap and `calc` functions are expensive, then this seems intuitive, otherwise maybe some other choice is needed. If both are going to be used, care will be needed to have the same distinction through the whole code base (in case this isn't already done)
- I'm not sold on the necessity of including `snap` in function names in the `snap` module, e.g. I think I prefer `from ...snap import ...descriptors... as ...snap_descriptors...` vs the existing `from ...snap import ...snap_drescriptors...`. This is a matter of taste and consistency with the rest of the modules is more important.
- Using big static arrays of numbers in the tests is both hard to understand and fragile to maintain. Better would be to write the code for generating some arrays such that they are forced to obey some features (symmetry, diagonality, or maybe clearly a perfect bulk fcc, or whatever) and then test for some property (e.g. contrasting (non)-uniformity of per-atom descriptors in bulk vs vacancy structures, or similar).

File: structuretoolkit/analyse/snap.py
Comment (jan-janssen) [reply structuretoolkit/analyse/snap.py]: The increment is `2` rather than `1`, so just starting at `j1` is not sufficient.

File: structuretoolkit/analyse/snap.py
Comment (jan-janssen) [reply structuretoolkit/analyse/snap.py:173]: The increment is `2` rather than `1`, so just starting at `j1` is not sufficient.

File: structuretoolkit/analyse/snap.py
Comment (jan-janssen) [reply structuretoolkit/analyse/snap.py:474]: The assert statement is a way to fail early. If users run their python code with `-O` to improve performance they do not get this nice error message but rather a more cryptic one when we try to read from the LAMMPS memory. So from my perspective it is a hint when things go wrong rather than a try/except condition.

**Comment by jan-janssen** [comment]
> * Per my original review comment, the functions would benefit from type hints. Additionally, some of the docstring type hints can be expanded, e.g. is `atom_types (list):` actually `atom_types (list[str]):`

I agree that we should add type hints to the whole `structuretoolkit` package as well as `atomistics` but I leave this to a separate pull request for now. 

> * I like that the function names have verbs in them, but the difference between `calc` and `get` is implicit. If `get` functions are cheap and `calc` functions are expensive, then this seems intuitive, otherwise maybe some other choice is needed. If both are going to be used, care will be needed to have the same distinction through the whole code base (in case this isn't already done)

While the initial intention was to highlight computationally expensive functions, I agree this is inconsistent with the rest of the package so I renamed the function to always use get rather than calc. 

> * I'm not sold on the necessity of including `snap` in function names in the `snap` module, e.g. I think I prefer `from ...snap import ...descriptors... as ...snap_descriptors...` vs the existing `from ...snap import ...snap_drescriptors...`. This is a matter of taste and consistency with the rest of the modules is more important.

At the moment the users are intended to use the function directly from the structure toolkit module: 
```python
from ase.build import bulk
import structuretookit as stk
stk.analyse.calc_snap_descriptors_per_atom(structure=bulk("Cu", cubic=True), atom_types=['Cu'])
```
This usage is also demonstrated in the tests in `TestSNAP`.

> * Using big static arrays of numbers in the tests is both hard to understand and fragile to maintain. Better would be to write the code for generating some arrays such that they are forced to obey some features (symmetry, diagonality, or maybe clearly a perfect bulk fcc, or whatever) and then test for some property (e.g. contrasting (non)-uniformity of per-atom descriptors in bulk vs vacancy structures, or similar).

The arrays only store the descriptor for a single atom while the structures use multiple atoms. But I am a bit reluctant to re-implement the SNAP descriptor in python just to validate the test.

File: structuretoolkit/analyse/snap.py
Comment (jan-janssen) [reply structuretoolkit/analyse/snap.py]: So you suggest to move the if statement one level up? I am not convinced that his improves the readability of the code.

Comment (liamhuber) [reply structuretoolkit/analyse/snap.py]: I'm saying replace j1-j2 and the if with a simple start=j2%2. Don't loop over stuff you're going to skip. This isn't about readability, it's about shortening the for loop.

File: structuretoolkit/analyse/snap.py
Comment (jan-janssen) [reply structuretoolkit/analyse/snap.py:541]: I added a corresponding comment.

File: structuretoolkit/analyse/snap.py
Comment (liamhuber) [reply structuretoolkit/analyse/snap.py]: I'm wrong about the nod, yes, moving the if our should do it

Comment (liamhuber) [reply structuretoolkit/analyse/snap.py]: Actually it's probably both, pull up j2<j1 as the if, then start at j2%2. But I'm on mobile and need to get some kids breakfast so I'll have to leave the details to you. No wasted loop space please though

Comment (jan-janssen) [reply structuretoolkit/analyse/snap.py]: Done

**Review by liamhuber — APPROVED** [review]
IMO type hinting can be implemented one module at a time, so I suggest going for it here and now to start the ball moving rather than delaying it to a future PR; I also don't quite follow your argument re the arrays, but this was a "tests better" rather than a "needs tests" comment so I'm not too stressed about it. Otherwise lgtm; the if clause replacement is very elegantly expressed btw.

**Pull request merged**
